src/maxi.c

```diff
@@ -0,0 +1,29 @@
+#include "MsCoreUtils.h"
+
+#include <R.h>
+#include <Rinternals.h>
+#include <math.h>
+
+
+/**
+ * Determine the maximal non-missing value returning NA if length(x) is
+ * 0 or if all values are missing.
+ */
+SEXP C_maxi(SEXP x) {
+    SEXP r;
+    const R_len_t n = XLENGTH(x);
+    PROTECT(r=allocVector(REALSXP, 1));
+    double* rp = REAL(r);
+    double* xp = REAL(x);
+    *rp = NA_REAL;
+
+    for (int i = 0; i < n; i++) {
+        if (!ISNA(xp[i])) {
+            if (ISNA(*rp) || (!ISNA(*rp) && xp[i] > *rp))
```

Comment:
You could reduce the number of comparison inside the loop to reduce the runtime.
1. Set `rp` to `-Inf` to avoid test for `NA` inside the loop and check for finite outside the loop.
2. No need to test `xp` for `NA` because `NA > *rp` is always false.

e.g.
``` r
library(microbenchmark)
library(collapse)
#> collapse 1.9.0, see ?`collapse-package` or ?`collapse-documentation`
#> 
#> Attaching package: 'collapse'
#> The following object is masked from 'package:stats':
#> 
#>     D
library(inline)
devtools::load_all()
#> ℹ Loading MsCoreUtils

maxir <- function(x) {
    res <- max(x, na.rm = TRUE)
    if (is.finite(res))
        res
    else NA_real_
}

maxi2 <- cfunction(c(x = "double"), "
  if (!isReal(x))
    x = coerceVector(x, REALSXP);
  double* xp = REAL(x);
  const R_xlen_t n = XLENGTH(x);

  SEXP r = PROTECT(allocVector(REALSXP, 1));
  double* rp = REAL(r);
  *rp = R_NegInf;

  for (R_xlen_t i = n; i >= 0; i--)
    if (xp[i] > *rp)
        *rp = xp[i];

  if (!(R_FINITE(*rp)))
    *rp = NA_REAL;

  UNPROTECT(1);
  return(r);
")

x <- rnorm(10, mean = 100)
x[4] <- NA
microbenchmark(
    max(x, na.rm = TRUE),
    maxi(x),
    maxi2(x),
    maxir(x),
    check = "identical"
)
#> Unit: nanoseconds
#>                  expr  min     lq     mean median     uq     max neval cld
#>  max(x, na.rm = TRUE)  413  463.0   578.59  476.5  558.0    3684   100   a
#>               maxi(x) 1172 1234.5  1491.55 1275.0 1372.5   11934   100   a
#>              maxi2(x)  656  695.5   822.53  725.0  794.0    4956   100   a
#>              maxir(x) 1201 1245.0 37377.27 1289.5 1419.0 3595709   100   a

x <- rnorm(1000, mean = 100)
x[4] <- NA
microbenchmark(
    max(x, na.rm = TRUE),
    maxi(x),
    maxi2(x),
    maxir(x),
    check = "identical"
)
#> Unit: microseconds
#>                  expr    min     lq     mean  median      uq    max neval  cld
#>  max(x, na.rm = TRUE)  2.611  2.700  2.80259  2.7860  2.8605  3.867   100  b  
#>               maxi(x) 11.940 12.101 12.78183 12.6715 12.8435 23.387   100    d
#>              maxi2(x)  1.802  1.897  2.21136  2.0020  2.0830 16.639   100 a   
#>              maxir(x)  3.396  3.547  3.76781  3.6620  3.7415  9.897   100   c
```

<sup>Created on 2023-02-08 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Reply:
impressive @sgibb ! Only thing is that `maxi2(numeric())` will return `-Inf` instead of `NA_real_`. When looping from 0 to n instead it works (I guess the idea was that `x` was an increasingly sorted vector such as m/z are)